File: fsharp/AoC2021/Day03/Day03/Program.fs
open System
open System.IO

let args = Environment.GetCommandLineArgs()

if args.Length <> 2 then
    failwith "Usage: $0 <INPUT FILE>"

let lines = File.ReadLines args.[1] |> Seq.toList

let readings =
    List.map (fun line -> Convert.ToUInt16(line, 2)) lines

let len = List.head lines |> String.length

printfn $"The power consumption is {Solution.computePowerConsumption len readings}."
printfn $"The life support rating is {Solution.computeLifeSupportRating len readings}."

File: csharp/AoC2021/Day03/Day03/Program.cs
using Day03;

if (args.Length != 1)
{
    Console.Error.WriteLine("Usage: $0 <input file>");
    Environment.Exit(1);
}

var len = 0;
var values = File
    .ReadAllLines(args[0])
    .Select(line =>
    {
        len = line.Length;
        return Convert.ToUInt16(line, 2);
    })
    .ToList();

var (gamma, epsilon) = Solution.ComputeGammaAndEpsilonRate(values, len);
Console.WriteLine($"The gamma rate is {gamma}, and the epsilon rate is {epsilon}, " +
                  $"so the final answer is {gamma * epsilon}");

var rating = Solution.ComputeLifeSupportRating(values, len);
if (rating == null)
    Console.Error.WriteLine("No life support rating could be determined.");
else
    Console.WriteLine($"The life support rating is {rating}.");
